src/e3/testsuite/control.py

```diff
@@ -0,0 +1,251 @@
+from enum import Enum
+import os.path
+
+from e3.testsuite import logger
+from e3.testsuite.optfileparser import OptFileParse
+
+
+class TestControlKind(Enum):
+    """Control how to run (or not!) testcases."""
+
+    NONE = 0
+    """Run the test the regular way."""
+
+    SKIP = 1
+    """Do not run the testcase, setting it UNSUPPORTED."""
+
+    XFAIL = 2
+    """
+    Run the test the regular way. If its status is PASS, correct it to
+    XPASS. If it succeeds, correct it to XFAIL. Leave its status unchanged in
+    other cases.
+    """
+
+
+class TestControl(object):
+    """Control the execution and analysis of a testcase."""
+    def __init__(self, message=None, skip=False, xfail=False):
+        """
+        :param None|str message: Optional message to convey with the test
+            status.
+        :param bool skip: Whether to skip the test execution.
+        :param bool xfail: Whether we expect the test to fail. If the test
+            should be skipped, and xfailed, we consider it failed even though
+            it did not run.
+        """
+        self.skip = skip
+        self.xfail = xfail
+        self.message = message
+
+
+class TestControlCreator(object):
+    """Abstract class to create test controls."""
+
+    def create(self, driver):
+        """Create a TestControl instance for the given test driver.
+
+        Raise a ValueError exception if the configuration is invalid.
+
+        :param TestDriver driver: Test driver for which we must parse the
+            "control" configuration.
+        """
+        raise NotImplementedError
+
+
+class YAMLTestControlCreator(TestControlCreator):
+    """Create test controls from "test.yaml"'s "control" entries."""
+
+    def __init__(self, condition_env={}):
+        """
+        :param dict condition_env: Environment to pass to condition evaluation
+            in control entries.
+        """
+        self.condition_env = condition_env
+
+    def create(self, driver):
+        # If there is a "test.opt" file while the YAML "control" entry
+        # mechanism is in use, it probably means someone mistakenly wrote a
+        # test.opt file that will not be interpreted: be helpful and warn about
+        # it.
+        if os.path.exists(driver.test_dir("test.opt")):
+            logger.warning(
+                '{}: "test.opt" file found whereas only "control" entries are'
+                ' considered'.format(driver.test_env["test_name"])
+            )
+
+        # Read the configuration from the test environment's "control" key, if
+        # present.
+        default = TestControl()
+        try:
+            control = driver.test_env["control"]
+        except KeyError:
```

Comment:
control = driver.test_env.get("control", default)

Reply:
No, that’s not equivalent: when there isn’t a "control" entry, we want to *return* the default TestControl instance.